### ISSUE TITLE
Possible false positive with Dictionary<K,V> Values property

### ISSUE THREAD

**Issue by RodrigoSerafim** [blaeser/parallelchecker#1]
When you do read only access to a Dictionary<K,V> Enumerator in two tasks a Race condition is issued, even though they are Read-Read operations. This also happens for properties like Count.

Example to replicate false positive:

``` csharp
Dictionary<string, string> cities = new Dictionary<string, string>() {
    { "new york", "usa"},
    { "munich", "germany"},
    { "lisbon", "portugal"},
};

//use this in the foreach to see the warning disappear
var other = cities.Values;
           
bool f = false;
bool t = true;
System.Threading.Tasks.Parallel.Invoke(
    () =>
    {
        foreach (var c in cities.Values)
            if (c == "usa")
                f = true;
    },
    () =>
    {
        foreach (var c in cities.Values)
            if (c == "germany")
                t = false;
    }
    );
Console.WriteLine($"{t} , {f}");
```

If you transform the collection into a IEnumerable first outside the thread (see  `other`  variable in the example) and use that instead then the warning goes away. Obtaining the iterator should be a thread safe operation for IDictionary. Some operations seem to be correctly marked as read only/thread safe while others with the same semantics aren't.

Warning issued:

```
Warning	ParallelChecker	Issue: #0 Race condition on System.Collections.Generic.Dictionary<string, string>
  caused by Thread-unsafe call at "cities.Values" in Program.cs line 20
    caused by thread or task at "() => { foreach (var c in cities..." in Program.cs line 18
      caused by call ConsoleConcurrent.Program.Main(string[])
        caused by initial thread at "Main" in Program.cs line 5
  caused by Thread-unsafe call at "cities.Values" in Program.cs line 26
    caused by thread or task at "() => { foreach (var c in cities..." in Program.cs line 24
      caused by call ConsoleConcurrent.Program.Main(string[])
        caused by initial thread at "Main" in Program.cs line 5
```

**Comment by blaeser** [comment]
Thank you very much for this report. The checker's precision can be improved here in two aspects:
- Read-only operations can be safely done as long as there is no modifying operation (according to the specification of Dictionary class). Currently, all concurrent operations on instances of generally thread-unsafe classes like Dictionary are treated as a potential issue.
- The copying of iterator references (as described in your example) should lead to the same analysis results. 

I will take these two issues on the list to be fixed with the next release.

**Comment by cryolithic** [comment]
I'm also seeing a false positive when accessing dictionary values behind a lock.

`
...
lock(someObj)
{
  SomeFunc(key);
}


...
`

`
private void SomeFunc(TKey key)
{
  if(someDict.ContainsKey(key)) //ParallelChecker marks this as a race condition on the value
  {
    //do stuff
  }
}
`

I assume that's due to the lock being external to SomeFunc, regardless of it only being called from within the lock?
